### PR TITLE
Space - Cathy

### PR DESCRIPTION
**Pull request by cojenco**
# Assignment Submission: Hotel

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | Where to hold the reservations and how a hotel block should fit in. Which class should hold this information and which needed to be its own class.
What was a design decision you made that changed over time over the project? | At first I had two containers for my reservations. After mid-point review, I decided to remove one of the containers and only let each room hold its own reservations. Looking back, I would refactor that again.
What was a concept you gained clarity on, or a learning that you'd like to share? | All the little things and managing dependencies. Also breaking up methods into smaller methods. While breaking up methods and calling methods, then I also gained clarity on which classes should have an inheritance or composition relationship. 
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | For the DateRange #overlapping method, a nominal case was testing an obvious date range that entirely overlapped and a date range that did not overlap at all. 
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | For the DateRange #overlapping method, one of the edge case was testing a date range that only overlapped the check-in date or check-out date.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I did stick to the principal of writing the test first, red, green and refactor. I still need more practice on pseudocode.

### CODE REVIEW

File: lib/date_range.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      raise ArgumentError.new("Live in the present! Dates should not be prior to today.") if start_date < Date.today
+      raise ArgumentError.new("End date should not be earlier than or the same day as start date.")  if end_date <= start_date
+      
+      @start_date = start_date
+      @end_date = end_date
+    end
+
+    def count_nights
+      nights = (@end_date - @start_date).to_i
+      return nights
```
Comment (kaidamasaki) [inline lib/date_range.rb:20]: You don't need to assign to a variable before returning:

```suggestion
      return (@end_date - @start_date).to_i
```

File: lib/date_range.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    end
+
+    def include_date(date)
+      return true if date >= @start_date && date < @end_date
+      return false
```
Comment (kaidamasaki) [inline lib/date_range.rb:25]: You can directly return the boolean condition:

```suggestion
      return date >= @start_date && date < @end_date
```

File: lib/date_range.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+module Hotel
+  class DateRange
+    
+    attr_accessor :start_date, :end_date
+
+    # your library code should assume that it's receiving Date objects to start
+    def initialize(start_date, end_date)
+      raise ArgumentError.new("Please pass in Date class instances.") if !(start_date.is_a? Date) || !(end_date.is_a? Date)
+      raise ArgumentError.new("Live in the present! Dates should not be prior to today.") if start_date < Date.today
```
Comment (kaidamasaki) [inline lib/date_range.rb:11]: It's helpful to include the bad arguments in message of the `ArgumentError`:

```suggestion
      raise ArgumentError.new("Live in the present! Dates should not be prior to today.  Bad start_date: #{start_date}") if start_date < Date.today
```

File: lib/no_availability_error.rb [hunk 4]
```diff
@@ -0,0 +1,3 @@
+
+class NoAvailabilityError < StandardError
```
Comment (kaidamasaki) [inline lib/no_availability_error.rb:2]: Custom exception class!  :tada:

File: lib/reservation.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
+require_relative 'date_range'
+
+module Hotel
+  class Reservation
+    attr_reader :id, :start_date, :end_date, :date_range, :room_id, :block
+    attr_accessor :cost
+
+    @@next_id = 1
```
Comment (kaidamasaki) [inline lib/reservation.rb:8]: Generally we advise against using a class variable.  In this case it's not too bad but it would be preferable to have this stored in the `SystemCoordinator`.

File: lib/reservation.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+      @date_range = date_range
+      @start_date = date_range.start_date
+      @end_date = date_range.end_date
+    
+      @room_id = room_id
+      @cost = cost
+
+      @id = @@next_id
+      @@next_id += 1
+      @block = -1
```
Comment (kaidamasaki) [inline lib/reservation.rb:20]: It looks like you're using `-1` to signal that this isn't a block.  

Generally speaking you should use `nil` to indicate the absence of something:

```suggestion
      @block = nil
```

File: lib/room.rb [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+      price = reservation.date_range.count_nights * cost
+      return price
+    end
+
+    def is_available(given_range)   
+      return true if bookings.empty?
+
+      bookings.each do |item|
+        if item.block >= 0   #item.class == Hotel::Block 
+          return false if item.date_range.overlapping(given_range) && item.date_range.exactly_matching(given_range)== false #it is not an exact match #method
```
Comment (kaidamasaki) [inline lib/room.rb:30]: Instead of comparing to `false` you should use `!`:

```suggestion
          return false if item.date_range.overlapping(given_range) && !item.date_range.exactly_matching(given_range) #it is not an exact match #method
```

File: lib/system_coordinator.rb [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+    def build_rooms(room_quantity)
+      room_quantity.times do
+        @rooms << Hotel::Room.new(@rooms.length + 1)
+      end
+    end
+
+    def list_rooms
+      return rooms
+    end
```
Comment (kaidamasaki) [inline lib/system_coordinator.rb:26]: You don't need both a `list_rooms` method and an `attr_reader` on `rooms`, you should pick one.

File: lib/system_coordinator.rb [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+        end
+      end
+
+      return reservations_by_date
+    end
+
+
+    def find_reservations_room_date(room_id, date_range)
+      selected_room = find_room(room_id)
+      reservations_room_date = selected_room.bookings.reject{|reservation|false == reservation.date_range.overlapping(date_range)}
```
Comment (kaidamasaki) [inline lib/system_coordinator.rb:44]: This can be simplified using `select`:

```suggestion
      reservations_room_date = selected_room.bookings.select {|reservation| reservation.date_range.overlapping(date_range)}
```

File: lib/system_coordinator.rb [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+        else
+          status = true
+          room.bookings.each do |item|
+            if item.block >= 0  #item.class == Hotel::Block 
+              status = false if item.date_range.overlapping(given_range) && false == item.date_range.exactly_matching(given_range) #it is not an exact match #method
+            elsif item.block < 0  #item.class == Hotel::Reservation
+              status = false if item.date_range.overlapping(given_range)
+            end
+          end
+          available_rooms << room if true == status
```
Comment (kaidamasaki) [inline lib/system_coordinator.rb:77]: You don't need to compare to `true` here since `status` is a boolean value:

```suggestion
          available_rooms << room if status
```

File: lib/system_coordinator.rb [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+require_relative 'date_range'
+require_relative 'reservation'
+require_relative 'block'
+require_relative 'no_availability_error'
+
+module Hotel
+  class SystemCoordinator    
+    attr_reader :rooms
+
+    @@next_block = 1
```
Comment (kaidamasaki) [inline lib/system_coordinator.rb:11]: This definitely doesn't need to be a class variable since there will only be one `SystemCoordinator` instantiated.

File: lib/system_coordinator.rb [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+    end
+
+
+    def make_specific_block(date_range, roomid_array, cost)
+      room_array = roomid_array.map{|id| find_room(id)} 
+      raise ArgumentError.new "Maximum 5 rooms for a hotel block." if room_array.length > 5 || room_array.length == 0
+      rooms_in_block = []
+
+      room_array.each do |room|
+        if room.is_available(date_range) == false
```
Comment (kaidamasaki) [inline lib/system_coordinator.rb:115]: This can be clarified using an `unless`:

```suggestion
        unless room.is_available(date_range)
```

**Comment by kaidamasaki** [comment]
# Hotel






## Section 1: Major Learning Goals



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
Practices SRP by having at least two separate classes with distinct responsibilities, and test files for these two classes | ✔️
Overall, demonstrates understanding instance variables vs. local variables. (There aren't unnecessarily too many instance variables, when it should be a local variable) | ✔️
For each test file, tests demonstrate an understanding of instantiating objects properly, and using Arrange-Act-Assert | ✔️
Practices pseudocode and TDD, and reflected on it by filling out the reflection questions | ✔️
Practices git with at least 15 small commits and meaningful commit messages | ✔️

## Section 2: Code Review and Testing Requirements

| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
There is a class that represents a reservation, and a second class that holds/manages a collection of reservations through composition (instance variable) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date is complex logic that is separated into method(s) (and potentially class(es)) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date has unit tests | ✔️
All of the given tests run and pass | ✔️
A test coverage tool is installed and used, and shows 95% test coverage | ✔️ :tada: 100%! :tada:

## Section 3: Feature Requirements

| Feature Requirement: There is a method that... | yes/no |
| --- | --- |
gives back a list of rooms, and it's tested | ✔️
creates a specific reservation for a room for a given date range, and it has nominal test cases | ✔️
creates a specific reservation for a room for a given date range, and it tests an edge case, such as no available room, or invalid date range | ✔️
gives back a list of reservations on a given date. Its tests include a test that makes several reservations for a given date | ✔️ (Tests on with multiple on the same day are borderline.)
calculates the total price for a reservation | ✔️
gives back a list of available rooms for a given date range, and it has nominal test cases | ✔️
gives back a list of available rooms for a given date range, and it has edge test cases, such as no available room, or invalid date range | (No edge case test.)
creates a block of rooms | ✔️
reserves a room from a block | (Not implemented.)

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ total in all sections | ✔️
| Yellow (Approaches Standards) | 9-13 total in all sections |
| Red (Not at Standard) | 0-8 total in all sections, or assignment is breaking/doesn’t run with less than 5 minutes of debugging |

### Additional Feedback



Great work overall! You've built your first project with minimal starting code. This represents an incredible milestone in your journey, and you should be proud of yourself!

I am particularly impressed by the way that you went about writing your tests.  You had really strong set up and assertion checks throughout your project.


It might make sense to re-think how you implemented blocks.  The way you have it you just make a series of unrelated reservations.  Once you've reserved a block of rooms you should be able to reserve those rooms _through_ that block, but not outside of it.  (They aren't just a bulk purchase of reservations with a discount.)


This project was extremely difficult though and just completing the basic functionality of blocks is a victory!


## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever |
| Descriptive/Readable |
| Concise |
| Logical/Organized | ✅

**Review by kaidamasaki — COMMENTED** [review]
Great job!  Here are some ways you can tidy up and clarify your code. 😄
